chasm/tree.go

```diff
@@ -344,6 +354,183 @@ func (n *Node) setSerializedNode(
 	childNode.setSerializedNode(nodePath[1:], serializedNode)
 }
 
+// serialize sets or updates serializedValue field of the node n with serialized value.
+func (n *Node) serialize() error {
+	switch n.serializedNode.GetMetadata().GetAttributes().(type) {
+	case *persistencespb.ChasmNodeMetadata_ComponentAttributes:
+		return n.serializeComponentNode()
+	case *persistencespb.ChasmNodeMetadata_DataAttributes:
+		return n.serializeDataNode()
+	case *persistencespb.ChasmNodeMetadata_CollectionAttributes:
+		panic("not implemented")
+	case *persistencespb.ChasmNodeMetadata_PointerAttributes:
+		panic("not implemented")
+	default:
+		return serviceerror.NewInternal("unknown node type")
+	}
+}
+
+func (n *Node) serializeComponentNode() error {
+	nodeValueT := reflect.TypeOf(n.value)
+	nodeValueV := reflect.ValueOf(n.value)
+
+	protoMessageFound := false
+	// TODO: consider using walker pattern to unify walking over reflected fields.
+	for i := 0; i < nodeValueT.Elem().NumField(); i++ {
+		fieldV := nodeValueV.Elem().Field(i)
+		fieldT := fieldV.Type()
+		if !fieldT.AssignableTo(protoMessageT) {
+			continue
+		}
+
+		if fieldV.IsNil() {
+			continue
+		}
```

Comment:
just to confirm,  this means component state should never be nil right?

Reply:
It can be `nil` and it indicates that `Data` field needs to be cleared. I am changing this part here.